pkg/virtualgateway/members_finalizer.go

```diff
@@ -0,0 +1,84 @@
+package virtualgateway
+
+import (
+	"context"
+	"fmt"
+	appmesh "github.com/aws/aws-app-mesh-controller-for-k8s/apis/appmesh/v1beta2"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/pkg/runtime"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strings"
+	"time"
+)
+
+const (
+	pendingMembersFinalizerEvaluateInterval = 60 * time.Second
+)
+
+type MembersFinalizer interface {
+	Finalize(ctx context.Context, vg *appmesh.VirtualGateway) error
+}
+
+// +kubebuilder:rbac:groups="",resources=events,verbs=get;list;watch;create;update;patch;delete
+
+func NewPendingMembersFinalizer(k8sClient client.Client, eventRecorder record.EventRecorder, log logr.Logger) MembersFinalizer {
+	return &pendingMembersFinalizer{
+		k8sClient:        k8sClient,
+		eventRecorder:    eventRecorder,
+		log:              log,
+		evaluateInterval: pendingMembersFinalizerEvaluateInterval,
+	}
+}
+
+// pendingMembersFinalizer is a MembersFinalizer that will pend virtualGateway deletion until all virtualGateway members are deleted.
+type pendingMembersFinalizer struct {
+	k8sClient     client.Client
+	eventRecorder record.EventRecorder
+	log           logr.Logger
+
+	evaluateInterval time.Duration
+}
+
+func (m *pendingMembersFinalizer) Finalize(ctx context.Context, vg *appmesh.VirtualGateway) error {
+	grMembers, err := m.findGatewayRouteMembers(ctx, vg)
```

Comment:
why do we need to find gateway routes from Gateway?

Reply:
This will block VirtualGateway's deletion if there are any GatewayRoutes attached to it.